upstairs/src/volume.rs

```diff
@@ -3490,55 +3510,116 @@ mod test {
         // A valid replacement VCR is provided with only one target being
         // different.
         // Test all three targets for replacement.
-        for cid in 0..3 {
-            test_volume_replace_cid(cid).unwrap();
+        // Test with 1, 2, and 3 sub_volumes.
+        // Test with the difference being in each sub_volume.
+        // Test both with and without a read_only_parent.
+        for sv in 1..4 {
+            for sv_changed in 0..sv {
+                for cid in 0..3 {
+                    test_volume_replace_inner(sv, sv_changed, cid, false)
+                        .unwrap();
+                    test_volume_replace_inner(sv, sv_changed, cid, true)
+                        .unwrap();
+                }
+            }
         }
     }
 
-    fn test_volume_replace_cid(cid: usize) -> Result<()> {
+    // A basic replacement of a target in a sub_volume, with options to
+    // create multiple sub_volumes, and to select which sub_volume and specific
+    // target to be different.
+    //
+    // sv_count: The number of sub volumes we create in each volume.
+    // sv_changed: The index (zero based) of the sub volume where we
+    // want the target to be different.
+    // cid: The client ID where the change will happen, which is the
+    // same as the index in the crucible opts target array.
+    // add_rop: True to create a read_only_parent, false for None
+    fn test_volume_replace_inner(
+        sv_count: usize,
+        sv_changed: usize,
+        cid: usize,
+        add_rop: bool,
```

Comment:
Nit: `add_rop` is confusing, because we don't change the ROP state.  Maybe `with_rop`?

Reply:
Also, _should_ we test removing the read-only parent?  We could use a three-variant enum in these tests, e.g.

```rust
enum ReadOnlyParentMode {
    Absent,
    Present,
    Remove,
}
```

Reply:
Since the loops are doing all the work here, we could even do:
```
ReadOnlyParentMode {
    Both,
    OnlyOriginal,
    OnlyReplacement,
    Neither,
}

The `OnlyReplacement` should fail 100% of the time.

Reply:
I did this in 2032eea3